gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +439,83 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
     return null;
   }
 
+  @VisibleForTesting
+  boolean isGoogleS2AEnabled() {
+    String S2AEnv = envProvider.getenv(S2A_ENV_ENABLE_USE_S2A);
+    boolean isS2AEnv = Boolean.parseBoolean(S2AEnv);
+    if (isS2AEnv) {
+      return true;
+    }
+    return false;
+  }
+
+  @VisibleForTesting
+  boolean shouldUseS2A() {
+    // If EXPERIMENTAL_GOOGLE_API_USE_S2A is not set to true, skip S2A.
+    if (!isGoogleS2AEnabled()) {
+      return false;
+    }
+
+    // If {@link mtlsEndpoint} is not set, skip S2A. S2A is also skipped when there is endpoint
+    // override. Endpoint override is respected when the {@link endpoint} is resolved via AIP#4114,
+    // see EndpointContext.java
+    if (endpoint != mtlsEndpoint) {
+      return false;
+    }
+
+    // mTLS via S2A is not supported in any universe other than googleapis.com.
+    if (!endpoint.contains(Credentials.GOOGLE_DEFAULT_UNIVERSE)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @VisibleForTesting
+  ChannelCredentials createMtlsToS2AChannelCredentials() throws IOException {
+    if (!isOnComputeEngine()) {
```

Comment:
```isOnComputeEngine``` is to tell whether we are on GCE/GKE. So it would return false if we were on Serverless. I'm pretty sure the function was added just for the DirectPath logic in this file. 

```createMtlsToS2AChannelCredentials``` gets called if we decide we ```shouldUseS2A```.

Reply:
I think we probably should use the other version, which tells whether we are on GCP in general. 
In the future Serverless will also have mtls-mds, and we don't want to have to comeback to update this.

Or do we need this check? I believe we only get to here, if we got a response from autoconfig endpoint with the s2a addresses, which already means we are on gcp?

Reply:
There isn't a utility here to see onGcp. 

But you are right, we only get here if autoconfig returns s2a addresses, which means we are on GCP. I've removed this call to isComputeEngine in the latest commit.